### PR TITLE
Adding SDK for QnAMaker v4

### PR DESCRIPTION
**Pull request by ghost**
Adding SDK for QnAMaker v4
Rest Spec: https://github.com/Azure/azure-rest-api-specs/blob/master/specification/cognitiveservices/data-plane/QnAMaker/stable/v4.0/QnAMaker.json
1 small change in Spec since Endpoint field was clashing. I have submitted a PR for that here:
https://github.com/Azure/azure-rest-api-specs/pull/4548

## Description

---

This checklist is used to make sure that common guidelines for a pull request are followed.
- [x] Please add REST spec PR link to the SDK PR
- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md).**
- [x] **The pull request does not introduce [breaking changes](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/Documentation/breaking-changes.md).**

### [General Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).

### [Testing Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.

### [SDK Generation Guidelines](https://github.com/Azure/azure-sdk-for-net/blob/AutoRest/.github/CONTRIBUTING.md#sdk-generation-guidelines)
- [x] If an SDK is being regenerated based on a new swagger spec, a link to the pull request containing these swagger spec changes has been included above.
- [x] The generate.cmd file for the SDK has been updated with the version of AutoRest, as well as the commitid of your swagger spec or link to the swagger spec, used to generate the code.
- [x] The `*.csproj` and `AssemblyInfo.cs` files have been updated with the new version of the SDK.

### CODE REVIEW

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj [hunk 1]
```diff
@@ -0,0 +1,25 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <!-- Please do not move/edit code below this line -->
+  <Import Project="$([MSBuild]::GetPathOfFileAbove('AzSdk.reference.props'))" />
+  <!-- Please do not move/edit code above this line -->
+
+  <PropertyGroup>
+    <PackageId>Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker</PackageId>
+    <Description>This client library provides access to the Microsoft Cognitive Services QnAMaker APIs.</Description>
+    <Version>2.2.0-preview</Version>
```
Comment (yangyuan) [inline src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj]: Why starts with 2.2.0? not something like 0.1.0 or 1.0.0?

**Review by yangyuan — COMMENTED** [review]
Look all good except the version choice.

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj (hunk 1 above)
Comment (bingisbestest) [reply src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj]: used the same as face sdk. Fixed.

**Review by yangyuan — APPROVED** [review]

**Comment by bingisbestest** [comment]
Who can merge this?

**Comment by yangyuan** [comment]
@bingisbestest The ADX team, just like the rest-specs repo.
They will do a more detail check to ensure this fits the Azure standard.

**Comment by bingisbestest** [comment]
@veronicagg  Please take a look

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj [hunk 2]
```diff
@@ -0,0 +1,25 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+  <!-- Please do not move/edit code below this line -->
+  <Import Project="$([MSBuild]::GetPathOfFileAbove('AzSdk.reference.props'))" />
+  <!-- Please do not move/edit code above this line -->
+
+  <PropertyGroup>
+    <PackageId>Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker</PackageId>
+    <Description>This client library provides access to the Microsoft Cognitive Services QnAMaker APIs.</Description>
+    <Version>1.0.0</Version>
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj]: Is the REST API stable or preview?

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/generate.ps1 [hunk 3]
```diff
@@ -0,0 +1 @@
+Start-AutoRestCodeGeneration -ResourceProvider "cognitiveservices/data-plane/QnAMaker" -AutoRestVersion "latest"
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/generate.ps1]: Was the code generated using this script? 
There should be 2 additional files generated, a .txt and a SdkInfo_*.cs please commit these files

**Review by dsgouda — CHANGES_REQUESTED** [review]

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj (hunk 2 above)
Comment (bingisbestest) [reply src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj]: stable its already in PROD

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+    <AssemblyName>Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker</AssemblyName>
+    <PackageTags>Microsoft Cognitive Services;Cognitive Services;Cognitive Services SDK;REST HTTP client;QnAMaker;QnAMaker API;QnAMaker SDK;Vision;netcore451511</PackageTags>
+    <PackageReleaseNotes>
+    <![CDATA[
```
[the remark this answers is not in the document]
Comment (bingisbestest) [reply src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj]: done

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/generate.ps1 (hunk 3 above)
Comment (bingisbestest) [reply src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/generate.ps1]: done

File: src/SDKs/_metadata/cognitiveservices_data-plane_QnAMaker.txt [hunk 5]
```diff
@@ -0,0 +1,14 @@
+﻿Installing AutoRest version: latest
+AutoRest installed successfully.
+Commencing code generation
+Generating CSharp code
+Executing AutoRest command
+cmd.exe /c autorest.cmd https://github.com/bingisbestest/azure-rest-api-specs/blob/qnav4-fixes/specification/cognitiveservices/data-plane/QnAMaker/readme.md --csharp --version=latest --reflect-api-versions --csharp-sdks-folder=D:\dev\sdk\sdkb\src\SDKs
```
Comment (dsgouda) [inline src/SDKs/_metadata/cognitiveservices_data-plane_QnAMaker.txt]: @bingisbestest it looks like the code has been generated with a REST spec in your fork. Code must be regenerated using REST spec merged in the Azure master branch. Please regenerate the code from the correct spec and update the PR

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/AzSdk.RP.props [hunk 6]
```diff
@@ -0,0 +1,7 @@
+<Project ToolsVersion="15.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <!--This file and it's contents are updated at build time moving or editing might result in build failure. Take due deligence while editing this file-->
+  <PropertyGroup>
+    <AzureApiTag />
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/AzSdk.RP.props]: Please run 
```
msbuild build.proj /t:build /p:Scope=SDKs\dataPlane\Knowledge\QnAMaker
```
and commit all the changes generated

**Review by dsgouda — CHANGES_REQUESTED** [review]

File: src/SDKs/_metadata/cognitiveservices_data-plane_QnAMaker.txt (hunk 5 above)
Comment (bingisbestest) [reply src/SDKs/_metadata/cognitiveservices_data-plane_QnAMaker.txt]: Yes was waiting for the change to go through. Updating.

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/AzSdk.RP.props (hunk 6 above)
Comment (bingisbestest) [reply src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/AzSdk.RP.props]: done + reran

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/Generated/Alterations.cs [hunk 7]
```diff
@@ -0,0 +1,339 @@
+// <auto-generated>
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License. See License.txt in the project root for
+// license information.
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/Generated/Alterations.cs]: The directory structure seems inconsistent.
This is how it should look.
Knowledge -> QnAMaker
Knowledge -> QnAMaker -> Generated
Knowledge -> QnAMaker -> Customization
Knowledge -> QnAMaker -> Properties
Knowledge -> QnAMaker.Tests
Knowledge -> QnAMaker.Tests->SessionRecords

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+    <AssemblyName>Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker</AssemblyName>
+    <PackageTags>Microsoft Cognitive Services;Cognitive Services;Cognitive Services SDK;REST HTTP client;QnAMaker;QnAMaker API;QnAMaker SDK;Vision;netcore451511</PackageTags>
+    <PackageReleaseNotes>
+    <![CDATA[ 
+        This is SDK for QnAMaker v4 API.	
+    ]]>
+    </PackageReleaseNotes>
+  </PropertyGroup>
+
+  <ItemGroup>
```
Comment (dsgouda) [inline src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj]: Please remove this ItemGroup

**Review by dsgouda — CHANGES_REQUESTED** [review]
Looks good for the most part, please address the remaining comments and this should be good

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj (hunk 8 above)
Comment (bingisbestest) [reply src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/QnAMaker/Microsoft.Azure.CognitiveServices.Knowledge.QnAMaker.csproj]: done

File: src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/Generated/Alterations.cs (hunk 7 above)
Comment (bingisbestest) [reply src/SDKs/CognitiveServices/dataPlane/Knowledge/QnAMaker/Generated/Alterations.cs]: done as discussed

**Review by dsgouda — APPROVED** [review]
LGTM

**Comment by dsgouda** [comment]
Will merge once CIs pass

**Comment by bingisbestest** [comment]
All CIs passed. Please merge.

**Pull request merged**